Add permission support for forms, lists and filters

Samuell1 opened the pull request:
Resolves #1837

Samuell1 committed:
Add Form permissions support

Samuell1 committed:
Add List permissions support

Samuell1 committed:
Add Filter permission support

Commit:
Merge branch 'develop' into form-list-filters-permissions

Commit:
Update unit tests to allow auth acting

Commit:
Add unit tests for form permissions check

bennothommo commented:
@Samuell1 I've added in some unit tests for the form permission check for you. I'll need to do the same for the lists and filters too.

Commit:
Prevent test users from being persisted in DB

Commit:
Use SQLite for normal unit tests

Commit:
Fix code smells

LukeTowers commented:
The property should probably be `permissions` and require an array, just like menu items: https://octobercms.com/docs/plugin/registration#navigation-menus

Commit:
Allow array syntax for form field permissions

bennothommo commented:
@LukeTowers I think the permissions check allows for either a single string or an array - I've changed it to `permissions` and allowed and checked for both types of definition in the tests.

(cc: @Samuell1)

LukeTowers commented on modules/backend/widgets/Filter.php:
```diff
@@ -561,6 +562,12 @@ public function addScopes(array $scopes)
 
+            // Check if user has permission to show this filter
+            $permission = array_get($config, 'permission');
+            if ($permission && !BackendAuth::getUser()->hasAccess($permission)) {
```
```suggestion
            if ($permission && !BackendAuth::getUser()->hasAnyAccess($permission)) {
```
Should be `hasAnyAccess()` instead of `hasAccess()`, see https://github.com/octobercms/october/blob/master/modules/backend/classes/NavigationManager.php#L520

LukeTowers replied:
Will need to be changed for all the permission checks (lists, forms, filters)

Samuell1 replied:
Can you then write it in single line if its single permission or then its needed to write it for everything as array?

LukeTowers replied:
@Samuell1 permissions can still be defined as a string (`permissions: backend.manage_users` for example) as hasAnyAccess just calls hasAccess($permission, $all = false) which calls hasPermission() which takes an array or a string as input: https://github.com/octobercms/library/blob/master/src/Auth/Models/User.php#L460

LukeTowers replied:
@Samuell1 the method called needs to be changed to `hasAnyAccess()` not just `hasAccess()`

Samuell1 committed:
Add other tests

Samuell1 committed:
Change model names

Samuell1 committed:
Change for all widgets

Samuell1 committed:
Update names

Samuell1 committed:
Add support for array of permissions

LukeTowers commented:
@Samuell1 looks like I was wrong, hasAnyAccess typehints an array: https://travis-ci.org/octobercms/october/jobs/582823639#L387

just change to `$permissions = is_string($permissions) ? [$permissions] : $permissions;` then check `if (!empty($permissions)` instead of `if ($permissions`

Samuell1 committed:
Revert "Add support for array of permissions"

This reverts commit a7f5ed2798d625df43f9bdbfd123945508f81951.

Commit:
Merge branch 'develop' into pr/Samuell1/4520-1

Commit:
Fix Lists column permissions unit tests

Commit:
Finish Filter widget tests

Commit:
Allow full testing suite to be enabled for certain tests

This will allow the new tests to take full advantage of migrations and authentication, but still keep the older tests running fast.

bennothommo commented:
@Samuell1 I've fixed the unit tests for you.

@LukeTowers this should be good to review.

LukeTowers commented on composer.json:
```diff
@@ -51,6 +51,7 @@
     },
     "autoload-dev": {
         "classmap": [
+            "tests/concerns/InteractsWithAuthentication.php",
```
Why does this have to be added to the composer file? Why can't it live with the rest of the test cases?

bennothommo replied:
@LukeTowers It's not a test case, it's a trait to enable emulating of an authenticated user. This was needed for the permissions tests. The reason for it being added there in Composer was because it is namespaced, but our tests folder in general is not currently namespaced.

LukeTowers commented on modules/backend/ServiceProvider.php:
```diff
@@ -88,6 +90,14 @@ protected function registerAssetBundles()
+     */
+    protected function registerFactories()
+    {
+        App::make(Factory::class)->load(__DIR__ . '/factories');
```
woah woah woah, why do we suddenly need factories?

bennothommo replied:
@LukeTowers This is used for generating the authenticated user fixture for testing the permissions. I would also be keen to use it down the line for some other unit tests as well.

LukeTowers replied:
Let's get @daftspunk's input on this

bennothommo replied:
@LukeTowers As per @daftspunk's recommendations, this will be moved to the tests folder and only available for tests.

bennothommo replied:
Further to the above, factories are now completely removed and replaced with a UserFixture class.

LukeTowers commented:
This needs to use `hasAnyAccess()` or `hasAccess($permissions, false)` instead to match the NavigationManager: https://github.com/octobercms/october/blob/master/modules/backend/classes/NavigationManager.php#L542. In this PR's current form, all permissions defined are required in order to be permitted. The existing NavigationManager `permissions` definition uses an `OR` check instead of an `AND` check.

Commit:
Refactor tests and required classes

- Removed factories
- Added static BackenUser fixture
- Removed full testing from TestCase
- Made PluginTestCase extend TestCase
- Fixed up tests

Samuell1 committed:
Allow OR in permissions

LukeTowers commented on modules/backend/widgets/Filter.php:
```diff
@@ -561,6 +562,12 @@ public function addScopes(array $scopes)
 
+            // Check if user has permissions to show this filter
+            $permissions = array_get($config, 'permissions');
+            if ($permissions && !BackendAuth::getUser()->hasAccess($permissions, false)) {
```
@Samuell1 what happens if $permissions is an empty array? Should this check be `!empty($permissions)` instead?

LukeTowers commented on modules/backend/widgets/Lists.php:
```diff
@@ -709,6 +710,10 @@ public function getColumns()
     {
+        if (!isset($this->allColumns[$column])) {
+            throw new ApplicationException('No definition for column ' . $column);
+        }
```
Could you provide some more background on the inclusion of this?

Samuell1 replied:
I think that was added by @bennothommo

bennothommo replied:
@LukeTowers I noticed that filters had an exception for a missing scope - see here:
https://github.com/octobercms/october/blob/33f5b3933f99c6dd6da93ab3bdd833d86dde317e/modules/backend/widgets/Filter.php#L890-L892
I copied the same idea to Lists to make it easy to test, however, it is arguably a breaking change. We could potentially make that method and the one in Lists return `null` if a scope/column does not exist instead.

LukeTowers replied:
I'm fine with leaving it in, was just curious on the background. Not really a breaking change because it would have thrown an error previously when trying to access an index that didn't exist.

Samuell1 committed:
Add additional empty check

LukeTowers commented:
@Samuell1 looks great, can you make a PR to octobercms/docs documenting the use of this for filters, lists, and forms?

LukeTowers commented on modules/backend/widgets/Form.php:
```diff
@@ -688,6 +689,12 @@ public function addFields(array $fields, $addToArea = null)
+            $permissions = array_get($config, 'permissions');
+            if (!empty($permissions) && !BackendAuth::getUser()->hasAccess($permissions, false)) {
+                continue;
+            }
```
@Samuell1 can you move this to be the first check of the loop? There's no need to make the fieldObj if we're not going to be using it.

LukeTowers commented on modules/backend/widgets/Filter.php:
```diff
@@ -561,6 +562,12 @@ public function addScopes(array $scopes)
+            $permissions = array_get($config, 'permissions');
+            if (!empty($permissions) && !BackendAuth::getUser()->hasAccess($permissions, false)) {
+                continue;
+            }
```
@Samuell1 can you move this to be the first check of the loop? There's no need to make the $scopeObj if we're not going to use it.

Samuell1 committed:
Move permission check to top

LukeTowers commented:
@Samuell1 looking good, added some change requests to the docs PR and then it should be good to go.
